VSM writeBlockMatrix

jbloom22 opened the pull request:
I'll add more tests, and I'm still considering whether to rip out or integrate WriteBlocksRDD for IRM. But given how similar this is to the latter, I'm ready for feedback on the new code.

cseed commented on src/main/scala/is/hail/distributedmatrix/BlockMatrix.scala:
```diff
@@ -752,36 +755,37 @@ class WriteBlocksRDD(irm: IndexedRowMatrix, path: String, gp: GridPartitioner) e
     var firstRowInNextBlock = 0L
-    var pi = 0 // parent partition index
+    var pi = 0
+    // parent partition index
```
Broken formatting

cseed commented on src/main/scala/is/hail/distributedmatrix/BlockMatrix.scala:
```diff
@@ -823,20 +827,153 @@ class WriteBlocksRDD(irm: IndexedRowMatrix, path: String, gp: GridPartitioner) e
+  }
+}
+
+class WriteBlocksRDD2(vsm: VariantSampleMatrix, path: String, parentPartitionStarts: Array[Long], f: () => java.lang.Double, ec: EvalContext, gp: GridPartitioner) extends RDD[Int](vsm.sparkContext, Nil) {
```
Hmm, taking a VSM might be a problem.  VSM isn't Serializable but RDDs need to be.  You might just take the things you need (which seem to include the sc as well as the variables below (rdd2, rowType, sampleIds, etc.))

cseed replied:
In fact, I'm not quite sure why this works.  Maybe because vsm is only a constructor and not a member because it isn't used in a method.  That rule always struck me as precarious.

cseed commented on src/test/scala/is/hail/variant/vsm/VSMSuite.scala:
```diff
@@ -448,4 +450,19 @@ class VSMSuite extends SparkSuite {
+
+      vsm.writeBlockMatrix(dirname, "g.GT.gt", blockSize)
+      
+      assert(BlockMatrix.read(hc, dirname).toLocalMatrix() === TestUtils.vdsToMatrixDouble(vsm).t)
```
This isn't going to fly since the first thing we're going to do after this is delete `vdsToMatrixDouble`.

cseed commented on src/main/scala/is/hail/distributedmatrix/BlockMatrix.scala:
```diff
@@ -823,20 +827,153 @@ class WriteBlocksRDD(irm: IndexedRowMatrix, path: String, gp: GridPartitioner) e
+    assert(rows == gp.nRows)
+    val nBlockRows = gp.nBlockRows
+        
+    val parts = Array.ofDim[Partition](nBlockRows)
```
Nuke ofDim.

cseed commented on src/main/scala/is/hail/distributedmatrix/BlockMatrix.scala:
```diff
@@ -823,20 +827,153 @@ class WriteBlocksRDD(irm: IndexedRowMatrix, path: String, gp: GridPartitioner) e
+  }
+}
+
+class WriteBlocksRDD2(vsm: VariantSampleMatrix, path: String, parentPartitionStarts: Array[Long], f: () => java.lang.Double, ec: EvalContext, gp: GridPartitioner) extends RDD[Int](vsm.sparkContext, Nil) {
```
This duplicates the indexed row version, as we discussed.  What's involved in nuking that?

cseed commented on src/main/scala/is/hail/distributedmatrix/KeyedIndexedRowMatrix.scala:
```diff
@@ -0,0 +1,29 @@
+import org.apache.spark.mllib.linalg.distributed.IndexedRowMatrix
+import org.json4s.jackson
+
+case class KeyedIndexedRowMatrix(irm: IndexedRowMatrix, keys: Option[Array[Any]], partStarts: Option[Array[Long]]) {
```
What's this about?  Is it temporary?  This seems like it's going in the wrong direction but maybe I don't understand.

jbloom22 replied:
I was experimenting with how to pass the partStarts through to write. I'm also not super happy with it.

jbloom22 commented:
I addressed comments apart from improving the tests on VSM. There are two options regarding plan for writing out a Spark IRM:
1) just delete it
2) keep it, pass partStarts through for efficiency, and cut down on code duplication

I tried the latter, creating KeyedIndexedRowMatrix as abstraction to handle both PCA and writing, and pushing common structure to an object WriteBlocksRDD.

cseed commented:
I thought we all agreed we wanted to delete it.  What's wrong with 1?  Deleting >>> writing new code every time.

cseed commented on src/main/scala/is/hail/distributedmatrix/BlockMatrix.scala:
```diff
@@ -719,69 +721,79 @@ private class BlockMatrixMultiplyRDD(l: BlockMatrix, r: BlockMatrix)
-    s"IndexedRowMatrix has ${gp.nRows} rows but RDD only has ${parentPartitionBoundaries.last} IndexedRows.")
-  
+class WriteBlocksRDD(path: String,
+  rdd2: RVD,
```
Call this `rvd`.  The one in VSM is going to get renamed.

cseed commented on src/main/scala/is/hail/distributedmatrix/BlockMatrix.scala:
```diff
@@ -719,69 +721,79 @@ private class BlockMatrixMultiplyRDD(l: BlockMatrix, r: BlockMatrix)
   private val d = digitsNeeded(gp.numPartitions)
-  private val sHadoopBc = irm.rows.sparkContext.broadcast(
-    new SerializableHadoopConfiguration(irm.rows.sparkContext.hadoopConfiguration))
+  private val sHadoopBc = rdd2.sparkContext.broadcast(
```
If you're going to take it as an argument, just call `sc.broadcast(...`.

cseed commented on src/main/scala/is/hail/distributedmatrix/BlockMatrix.scala:
```diff
@@ -719,69 +721,79 @@ private class BlockMatrixMultiplyRDD(l: BlockMatrix, r: BlockMatrix)
-  private val sHadoopBc = irm.rows.sparkContext.broadcast(
-    new SerializableHadoopConfiguration(irm.rows.sparkContext.hadoopConfiguration))
+  private val sHadoopBc = rdd2.sparkContext.broadcast(
+    new SerializableHadoopConfiguration(rdd2.sparkContext.hadoopConfiguration))
```
same.

cseed commented on src/main/scala/is/hail/distributedmatrix/BlockMatrix.scala:
```diff
@@ -719,69 +721,79 @@ private class BlockMatrixMultiplyRDD(l: BlockMatrix, r: BlockMatrix)
+  sc: SparkContext,
+  rowType: TStruct,
+  sampleIds: IndexedSeq[Annotation],
+  sampleAnnotations: IndexedSeq[Annotation],
```
You should the broadcasted versions to avoid serializing the original values.

cseed commented on src/main/scala/is/hail/distributedmatrix/BlockMatrix.scala:
```diff
@@ -795,55 +807,64 @@ class WriteBlocksRDD(irm: IndexedRowMatrix, path: String, gp: GridPartitioner) e
+          var j = 0
+          while (j < n) {
+            ec.set(3, sampleIdsBc.value(j))
+            ec.set(4, sampleAnnotationsBc.value(j))
```
You need `offset + j` in the sample annotations, too.  You might write this to be a bit safer: one counter for the offset in the block, another for the sample.

jbloom22 replied:
Good catch. I've gone the safer route, and adjusted the test so it would have caught this.

cseed commented on src/main/scala/is/hail/variant/VariantSampleMatrix.scala:
```diff
@@ -2687,15 +2685,51 @@ class VariantSampleMatrix(val hc: HailContext, val metadata: VSMMetadata,
     }
 
-    (someIf(getVariants,
+    val irm = new IndexedRowMatrix(indexedRows, partStarts.last, nSamples)
```
Unrelated, incidental changes are not free.  Try to balance the benefit (before and after here seem essentially the same to me) with the cost (it needs to be reviewed, there is danger to introduce bugs, increase changes of merge conflicts, etc.)

jbloom22 replied:
Roger that, I've removed them.

cseed commented on src/main/scala/is/hail/utils/richUtils/RichIndexedRowMatrix.scala:
```diff
@@ -40,18 +39,18 @@ object RichIndexedRowMatrix {
 }
 
-class RichIndexedRowMatrix(indexedRowMatrix: IndexedRowMatrix) {
+class RichIndexedRowMatrix(irm: IndexedRowMatrix) {
```
Same.  I often review the differences before I PR and remove inessential changes that got made during the development of the PR.

jbloom22 commented:
I addressed the comments and added a success indicator on write that's checked on read since these are often going to be big objects.
